### PR TITLE
SOLR-16759: Introducing logAll parameter in the feature logger

### PR DESCRIPTION
**Pull request by aruggero**
https://issues.apache.org/jira/browse/SOLR-16759

# Description

Currently,  there is no possibility of deciding how many features to log.
It would be useful to give the user the ability to decide if to print all the features of the store or only the features used by the model.
Suppose to execute a query with both logging and reranking, where the logging store is the same as the model store. 
Suppose the model store contains 200 features while the model uses just half of them.
The logging will always consider all the 200 features of the store. 
There is no possibility of printing only the 100 features used by the model.

# Solution

An external parameter _logAll_ has been added to the features transformer in order to give the user the possibility to choose how many features to log.
The new parameter has been added to:
- solr/modules/ltr/src/java/org/apache/solr/ltr/FeatureLogger.java
- solr/modules/ltr/src/java/org/apache/solr/ltr/CSVFeatureLogger.java

# Tests

Several tests have been added into org/apache/solr/ltr/feature/TestFeatureLogging.java
We add one test for each feature logger store - logAll parameter combination.
We checked that the set of features returned is the expected on (all the features in the store or only the model features).

# Checklist

Please review the following and check all that apply:

- [X] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [X] I have created a Jira issue and added the issue ID to my pull request title.
- [X] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [X] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [X] I have added tests for my changes.
- [X] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/modules/ltr/src/java/org/apache/solr/ltr/FeatureLogger.java [hunk 1]
```diff
@@ -33,9 +33,13 @@ public enum FeatureFormat {
 
   protected final FeatureFormat featureFormat;
 
-  protected FeatureLogger(String fvCacheName, FeatureFormat f) {
+  protected final Boolean extractAll;
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/FeatureLogger.java:36]: logAll also here?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java [hunk 2]
```diff
@@ -29,8 +29,11 @@ public class SolrQueryRequestContextUtils {
   /** key of the scoring queries in the request context */
   private static final String SCORING_QUERIES = LTR_PREFIX + "scoring_queries";
 
+  /** key of the isPrintingFeatures flag in the request context */
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java:32]: isLoggingFeatures?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/FeatureLogger.java (hunk 1 above)
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/FeatureLogger.java:36]: logAll also here?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java [hunk 3]
```diff
@@ -29,8 +29,11 @@ public class SolrQueryRequestContextUtils {
...
+  private static final String IS_LOGGING_FEATURES = LTR_PREFIX + "isLoggingFeatures";
+
   /** key of the isExtractingFeatures flag in the request context */
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java:35]: isLoggingAllFeatures?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java [hunk 4]
```diff
@@ -54,18 +57,22 @@ public static LTRScoringQuery[] getScoringQueries(SolrQueryRequest req) {
   }
 
   /** isExtractingFeatures flag accessors */
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java:59]: isLoggingFeatures?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java [hunk 5]
```diff
@@ -54,18 +57,22 @@ public static LTRScoringQuery[] getScoringQueries(SolrQueryRequest req) {
...
 
-  public static void clearIsExtractingFeatures(SolrQueryRequest req) {
-    req.getContext().put(IS_EXTRACTING_FEATURES, Boolean.FALSE);
+  public static boolean isLoggingFeatures(SolrQueryRequest req) {
+    return Boolean.TRUE.equals(req.getContext().get(IS_LOGGING_FEATURES));
   }
 
-  public static boolean isExtractingFeatures(SolrQueryRequest req) {
-    return Boolean.TRUE.equals(req.getContext().get(IS_EXTRACTING_FEATURES));
+  public static void logAllFeatures(SolrQueryRequest req, Boolean logAll) {
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java:68]: setLogAllFeatures?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java [hunk 6]
```diff
@@ -146,7 +153,7 @@ public DocTransformer create(String name, SolrParams localparams, SolrQueryReque
    *
    * @return a feature logger for the format specified.
    */
-  private FeatureLogger createFeatureLogger(String formatStr) {
+  private FeatureLogger createFeatureLogger(String formatStr, Boolean extractAll) {
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java:156]: logAll?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java [hunk 7]
```diff
@@ -156,7 +163,7 @@ private FeatureLogger createFeatureLogger(String formatStr) {
     if (fvCacheName == null) {
       throw new IllegalArgumentException("a fvCacheName must be configured");
     }
-    return new CSVFeatureLogger(fvCacheName, format, csvKeyValueDelimiter, csvFeatureSeparator);
+    return new CSVFeatureLogger(fvCacheName, format, extractAll, csvKeyValueDelimiter, csvFeatureSeparator);
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java:166]: logAll?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java (hunk 5 above)
Comment (alessandrobenedetti) [reply solr/modules/ltr/src/java/org/apache/solr/ltr/SolrQueryRequestContextUtils.java:68]: this one I am not sure, I think the name can be reasonable without the explicit set

**Review by alessandrobenedetti — APPROVED** [review]
I think this PR is now in a good state and ready for additional reviews!

File: solr/modules/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java [hunk 8]
```diff
@@ -71,8 +71,13 @@ public class LTRFeatureLoggerTransformerFactory extends TransformerFactory {
   // used inside fl to specify the feature store to use for the feature extraction
   private static final String FV_STORE = "store";
 
+  // used inside fl to specify to log (all|model only) features
+  private static final String FV_LOG_ALL = "logAll";
+
   private static String DEFAULT_LOGGING_MODEL_NAME = "logging-model";
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java:77]: Final?

File: solr/modules/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java [hunk 9]
```diff
@@ -71,8 +71,13 @@ public class LTRFeatureLoggerTransformerFactory extends TransformerFactory {
...
 
+  private static boolean DEFAULT_NO_RERANKING_LOGGING_ALL = true;
```
Comment (aruggero) [inline solr/modules/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java:79]: Final?

**Review by alessandrobenedetti — APPROVED** [review]
Proceeding with merge

**Pull request merged**